README.md
# response_x

[![pub package](https://img.shields.io/pub/v/response_x.svg)](https://pub.dartlang.org/packages/response_x)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

library for exchanging response [JSON](https://en.wikipedia.org/wiki/JSON) between server and client. it can be a useful tool for standardizing the way that a server communicates with its clients.

# Installing

### 1. Depend on it

Add this to your package's `pubspec.yaml` file:

```yaml
dependencies:
  response_x: 1.0.0
```

### 2. Install it

You can install packages from the command line:

with `pub`:

```
$ pub get
```

with `Flutter`:

```
$ flutter pub get
```

### 3. Import it

Now in your `Flutter` code, you can use:

```dart
import 'package:response_x/response_x.dart';
```

### 4. Getting Started

Import the library.

```dart
import 'package:response_x/response_x.dart';
```

```dart
final response = Response.success();
print(response.statusCode); // 200
print(response.message); // OK
print(response.data); // null
print(response.success); // true

final response = Response.failure();
print(response.statusCode); // 400
print(response.message); // failure
print(response.data); // null
print(response.success); // false
```

### 5. Example

```dart
import 'package:response_x/response_x.dart';
import 'package:http/http.dart' as http;

Future<Response> todos() async {
  final response = await http.get('https://jsonplaceholder.typicode.com/todos');
  if (response.statusCode == 200) {
    return Response.success(data: response.body, message: "todos fetched successfully");
  } else {
    return Response.error(message: 'there was an error fetching todos');
  }
}


void main() async {
  final response = await todos();
  if (response.success) {
    print(response.data);
  } else {
    print(response.message);
  }
}

```

# Maintainers
[@ravikovind](https://ravikovind.github.io/)


CHANGELOG.md
## 1.0.0

- Initial release

## 1.0.1
- Fix: Added example to README.md

## 1.0.2
- Fix: A example project was added to the package.
- Test: Added tests for the package.

test/response_x_test.dart
import 'package:test/test.dart';

import 'package:response_x/response_x.dart';

void main() {
  group('Response', () {
    test('Response.success', () {
      final response = Response.success();
      expect(response.success, true);
      expect(response.statusCode, 200);
      expect(response.message, 'ok');
      expect(response.data, null);
    });
    test('Response.redirect', () {
      final response = Response.redirect();
      expect(response.success, true);
      expect(response.statusCode, 302);
      expect(response.message, 'redirect');
      expect(response.data, null);
    });
    test('Response.failure', () {
      final response = Response.failure();
      expect(response.success, false);
      expect(response.statusCode, 400);
      expect(response.message, 'failure');
      expect(response.data, null);
    });
    test('Response.status', () {
      final listOfCodes = [
        100,
        200,
        201,
        202,
        203,
        204,
        205,
        206,
        207,
        208,
        226,
        300,
        301,
        302,
        303,
        304,
        305,
        306,
        307,
        308,
        400,
        401,
        402,
        403,
        404,
        405,
        406,
        407,
        408,
        409,
        410,
        411,
        412,
        413,
        414,
        415,
        416,
        417,
        418,
        421,
        422,
        423,
        424,
        425,
        426,
        428,
        429,
        431,
        451,
        500,
        501,
        502,
        503,
        504,
        505,
        506,
        507,
        508,
        510,
        511,
      ];

      for (final code in listOfCodes) {
        final response = Response.status(code);
        final success = code >= 200 && code < 400;
        final message = _kCodeToSuccessAndMessage[code]?[success]?.toString() ??
            'unable to get message for code $code';
        expect(response.success, success);
        expect(response.statusCode, code);
        expect(response.message, message);
        expect(response.data, null);
      }
    });
  });
}

const Map<int, Map<bool, String>> _kCodeToSuccessAndMessage = {
  100: {true: 'continue'},
  101: {true: 'switching protocols'},
  102: {true: 'processing'},
  103: {true: 'early hints'},
  200: {true: 'ok'},
  201: {true: 'created'},
  202: {true: 'accepted'},
  203: {true: 'non-authoritative information'},
  204: {true: 'no content'},
  205: {true: 'reset content'},
  206: {true: 'partial content'},
  207: {true: 'multi-status'},
  208: {true: 'already reported'},
  226: {true: 'im used'},
  300: {true: 'multiple choices'},
  301: {true: 'moved permanently'},
  302: {true: 'redirect'},
  303: {true: 'see other'},
  304: {true: 'not modified'},
  305: {true: 'use proxy'},
  306: {true: 'switch proxy'},
  307: {true: 'temporary redirect'},
  308: {true: 'permanent redirect'},
  400: {false: 'failure'},
  401: {false: 'unauthorized'},
  402: {false: 'payment required'},
  403: {false: 'forbidden'},
  404: {false: 'not found'},
  405: {false: 'method not allowed'},
  406: {false: 'not acceptable'},
  407: {false: 'proxy authentication required'},
  408: {false: 'request timeout'},
  409: {false: 'conflict'},
  410: {false: 'gone'},
  411: {false: 'length required'},
  412: {false: 'precondition failed'},
  413: {false: 'payload too large'},
  414: {false: 'uri too long'},
  415: {false: 'unsupported media type'},
  416: {false: 'range not satisfiable'},
  417: {false: 'expectation failed'},
  418: {false: 'i\'m a teapot'},
  421: {false: 'misdirected request'},
  422: {false: 'unprocessable entity'},
  423: {false: 'locked'},
  424: {false: 'failed dependency'},
  425: {false: 'too early'},
  426: {false: 'upgrade required'},
  428: {false: 'precondition required'},
  429: {false: 'too many requests'},
  431: {false: 'request header fields too large'},
  451: {false: 'unavailable for legal reasons'},
  500: {false: 'internal server error'},
  501: {false: 'not implemented'},
  502: {false: 'bad gateway'},
  503: {false: 'service unavailable'},
  504: {false: 'gateway timeout'},
  505: {false: 'http version not supported'},
  506: {false: 'variant also negotiates'},
  507: {false: 'insufficient storage'},
  508: {false: 'loop detected'},
  510: {false: 'not extended'},
  511: {false: 'network authentication required'},
};


example/test/example_test.dart
import 'package:example/example.dart' as example;

import 'package:test/test.dart';

Future<void> main() async {
  /// [resultTodos] is of type `Response`
  final resultTodos = await example.todos();

  test('Test Todos', () {
    expect(resultTodos.success, true);
    expect(resultTodos.statusCode, 200);
    expect(resultTodos.message, 'Todos fetched successfully');
    expect(resultTodos.data, isNotNull);
    expect(resultTodos.data, TypeMatcher<List<example.TODO>>());
  });

  /// [resultTodoById] is of type `Response`
  final resultTodoById = await example.todoById(1);

  test('Test Todo by id: 1', () {
    final todo = resultTodoById.data as example.TODO;
    expect(resultTodoById.success, true);
    expect(resultTodoById.statusCode, 200);
    expect(resultTodoById.message, 'Todo fetched successfully');
    expect(todo, TypeMatcher<example.TODO>());
  });
}


example/lib/example.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:response_x/response_x.dart' as res;

Future<res.Response> todos() async {
  try {
    final uri = Uri.parse(kURI);
    final response = await http.get(uri);
    if (response.statusCode == 200) {
      final todos = List<TODO>.from(
        json.decode(response.body)?.map((x) => TODO.fromJson(x)) ?? <TODO>[],
      );
      return res.Response.success(
        data: todos,
        message: 'Todos fetched successfully',
      );
    } else {
      return res.Response.failure(
        message: 'There was an error fetching todos',
      );
    }
  } on Exception catch (_) {
    return res.Response.failure(
      message: 'There was an error fetching todos : $_',
    );
  }
}

Future<res.Response> todoById(int id) async {
  try {
    final uri = Uri.parse('$kURI/$id');
    final response = await http.get(uri);
    if (response.statusCode == 200) {
      final todo = TODO.fromJson(json.decode(response.body));
      return res.Response.success(
        data: todo,
        message: 'Todo fetched successfully',
      );
    } else {
      return res.Response.failure(
        message: 'There was an error fetching todo',
      );
    }
  } on Exception catch (_) {
    return res.Response.failure(
      message: 'There was an error fetching todo : $_',
    );
  }
}

const kBaseURL = 'https://jsonplaceholder.typicode.com';
const kTodosPath = '/todos';
const kURI = '$kBaseURL$kTodosPath';

class TODO {
  TODO({
    required this.id,
    required this.title,
    required this.completed,
  });

  final int id;
  final String title;
  final bool completed;

  factory TODO.fromJson(Map<String, dynamic> json) {
    final id = int.tryParse('${json['id']}') ?? 0;
    final title = json['title']?.toString() ?? 'Not Provided';
    final completed = '${json['completed']}' == 'true';
    return TODO(
      id: id,
      title: title,
      completed: completed,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'completed': completed,
      };

  @override
  String toString() => '${toJson()}';
}


lib/src/response_x.dart
import 'package:equatable/equatable.dart';

/// [Response] is class that represents a server response. class can be a useful tool for standardizing the way that a server communicates with its clients.
class Response extends Equatable {
  /// [Response] constructor. it takes [success], [statusCode], [message] and [data] as parameters.
  const Response({
    required this.success,
    required this.statusCode,
    required this.message,
    this.data,
  });

  ///success is a [bool] value indicating whether the server operation was successful or not. It can be used to determine whether to proceed with further processing or show an error message to the user.
  final bool success;

  /// statusCode is typically an [int] value that represents the HTTP status code returned by the server. Common status codes include 200 (OK), 400 (Bad Request), 401 (Unauthorized), 404 (Not Found), and 500 (Internal Server Error).
  final int statusCode;

  /// message is a [String] containing a human-readable message that provides more information about the response. For example, if the [success] property is false, the [message] property could contain an error message that describes what went wrong.
  final String message;

  /// data is what the server wants to return to the client. It's a nullable [Object], It could be a [Map], a [List], or a single value such as a [String], [int], or [bool]. also it can be [null].
  final Object? data;

  /// [Response.success] returns a [Response] with [success] set to true and [statusCode] set to 200.
  /// [Response.success] can use for returning successful responses.
  /// generally successful responses are returned with [statusCode] 2xx.
  factory Response.success({
    String message = 'ok',
    Object? data,
  }) =>
      Response(
        success: true,
        statusCode: 200,
        message: message,
        data: data,
      );

  /// [Response.redirect] returns a [Response] with [success] set to true and [statusCode] set to 302.
  /// [message] is set to the provided [message].
  /// [data] is set to the provided [data].
  /// [data] is optional.
  /// [Response.redirect] can use for returning redirect responses.
  /// generally redirect responses are returned with [statusCode] 3xx.

  factory Response.redirect({
    String message = 'redirect',
    Object? data,
    bool success = true,
  }) =>
      Response(
        success: success,
        statusCode: 302,
        message: message,
        data: data,
      );

  /// [Response.failure] returns a [Response] with [success] set to false and [statusCode] set to the provided [statusCode] starting with 4xx or 5xx.
  /// [message] is set to the provided [message].
  /// [data] is set to the provided [data].
  /// [Response.failure] can use for returning failure responses.
  /// generally failure responses are returned with [statusCode] 4xx and 5xx.
  factory Response.failure({
    int statusCode = 400,
    String message = 'failure',
    Object? data,
  }) =>
      Response(
        success: false,
        statusCode: statusCode,
        message: message,
        data: data,
      );

  /// [Response.status] returns a [Response] based on the provided [statusCode].
  ///
  /// Reference: [https://developer.mozilla.org/en-US/docs/Web/HTTP/Status]
  factory Response.status(int code) {
    final success = code >= 200 && code < 400;
    final message = _kCodeToSuccessAndMessage[code]?[success]?.toString() ??
        'unable to get message for code $code';
    return Response(
      success: success,
      statusCode: code,
      message: message,
    );
  }

  /// [Response.json] returns a [Response].
  /// it's used overriding only [data] of the [Response].
  Response json(Object? data) => Response(
        success: success,
        statusCode: statusCode,
        message: message,
        data: data ?? this.data,
      );

  /// [Response.copyWith] returns a [Response] with the provided [success], [statusCode], [message] and [data].
  /// [success] is optional.
  /// [statusCode] is optional.
  /// [message] is optional.
  /// [data] is optional.
  /// it is useful for overriding the [success], [statusCode], [message] and [data] of the [Response].
  Response copyWith({
    bool? success,
    int? statusCode,
    String? message,
    Object? data,
  }) =>
      Response(
        success: success ?? this.success,
        statusCode: statusCode ?? this.statusCode,
        message: message ?? this.message,
        data: data ?? this.data,
      );

  /// [Response.fromJson] returns a [Response] from the provided [json] map of [String] and [dynamic].
  factory Response.fromJson(Map<String, dynamic> json) {
    final data = json['data'];
    final message = json['message']?.toString() ?? 'internal server error';
    final statusCode = int.tryParse('${json['statusCode']}') ?? 500;
    final success = json['success']?.toString() == 'true';
    return Response(
      success: success,
      statusCode: statusCode,
      message: message,
      data: data,
    );
  }

  /// [Response.toJson] returns a JSON [Map] of representation of the [Response].
  Map<String, dynamic> toJson() => {
        'success': success,
        'code': statusCode,
        'message': message,
        'data': data,
      };

  @override
  List<Object?> get props => [success, statusCode, message, data];

  @override
  String toString() =>
      'Response { success: $success, statusCode: $statusCode, message: $message, data: $data }';
}

/// [_kCodeToSuccessAndMessage] is a [Map] of [int] and [Map] of [bool] and [String].
/// [_kCodeToSuccessAndMessage] contains the [statusCode] and the [success] and the [message] of the response.
const Map<int, Map<bool, String>> _kCodeToSuccessAndMessage = {
  200: {true: 'ok'},
  201: {true: 'created'},
  202: {true: 'accepted'},
  203: {true: 'non-authoritative information'},
  204: {true: 'no content'},
  205: {true: 'reset content'},
  206: {true: 'partial content'},
  207: {true: 'multi-status'},
  208: {true: 'already reported'},
  226: {true: 'im used'},
  300: {true: 'multiple choices'},
  301: {true: 'moved permanently'},
  302: {true: 'redirect'},
  303: {true: 'see other'},
  304: {true: 'not modified'},
  305: {true: 'use proxy'},
  306: {true: 'switch proxy'},
  307: {true: 'temporary redirect'},
  308: {true: 'permanent redirect'},
  400: {false: 'failure'},
  401: {false: 'unauthorized'},
  402: {false: 'payment required'},
  403: {false: 'forbidden'},
  404: {false: 'not found'},
  405: {false: 'method not allowed'},
  406: {false: 'not acceptable'},
  407: {false: 'proxy authentication required'},
  408: {false: 'request timeout'},
  409: {false: 'conflict'},
  410: {false: 'gone'},
  411: {false: 'length required'},
  412: {false: 'precondition failed'},
  413: {false: 'payload too large'},
  414: {false: 'uri too long'},
  415: {false: 'unsupported media type'},
  416: {false: 'range not satisfiable'},
  417: {false: 'expectation failed'},
  418: {false: 'i\'m a teapot'},
  421: {false: 'misdirected request'},
  422: {false: 'unprocessable entity'},
  423: {false: 'locked'},
  424: {false: 'failed dependency'},
  425: {false: 'too early'},
  426: {false: 'upgrade required'},
  428: {false: 'precondition required'},
  429: {false: 'too many requests'},
  431: {false: 'request header fields too large'},
  451: {false: 'unavailable for legal reasons'},
  500: {false: 'internal server error'},
  501: {false: 'not implemented'},
  502: {false: 'bad gateway'},
  503: {false: 'service unavailable'},
  504: {false: 'gateway timeout'},
  505: {false: 'http version not supported'},
  506: {false: 'variant also negotiates'},
  507: {false: 'insufficient storage'},
  508: {false: 'loop detected'},
  510: {false: 'not extended'},
  511: {false: 'network authentication required'},
};
